documentation/design_docs/design/infra-resource.adoc

```diff
@@ -0,0 +1,134 @@
+== why
+
+Changing configuration for the standard or brokered address space today involves changing a
+`resource definition` configmap, which in turn references templates (no schema validation and
+checking of values for instance). The underlying openshift templates which are stored in a configmap
+as key-value pairs and can only be changed before deploying an address space or it will have to be
+recreated, unless you happen to know exactly what component to change.
+
+The settings that can be tuned and will have an effect when changed is not clear to spot, and
+are scattered around in the template.
+
+Finally, there is at present no mechanism for the service admin to upgrade the infrastructure
+between different versions of EnMasse and even for changing configuration run-time and make sure the
+appropriate components are changed.
+
+== what/how
+
+This proposal attempts to address these issues by defining a format for infrastructure configuration
+that clearly states the configuration settings that are available for tuning. Multiple versions of
+the configuration may exist in the system, and a particluar instance of a configuration is mapped to
+one of the configuration versions.
+
+Upgrades are handled by updating the version of the instance of the infrastructure, in which case a
+controller will take steps to migrate the infra from one version to the next.
+
+Defining a syntax for configuring infrastructure makes the operational life of EnMasse simpler.
+It also allows us to decouple definition of the infrastructure from the address space(s) that the
+infrastructure serves, which is a feature that is desired for kafka/strimzi support.
+
+A `StandardInfraTemplate` resource defines an infrastructure definition of admin, router and broker
+components. All supported settings for these components can be tuned through this resource. More
+settings that is desired to be configurable can be added. The resource is versioned in the name, and
+as a property.
+
+
+```
+apiVersion: enmasse.io/v1alpha1
+kind: StandardInfraTemplate
+metadata:
+  name: standard-infra-0.21.2
+  namespace: somewhere
+spec:
+  version: 0.21.2
+  router:
+    replicas: 1
+    linkCapacity: 1000
+    allowDurableSubscriptions: false
+    resources:
+      memory: 512Mi
+  broker:
+    addressFullPolicy: FAIL
+    resources:
+      memory: 512Mi
+      disk: 2Gi
+  admin:
+    resources:
+      memory: 512Mi
+```
+
+A `BrokeredInfraTemplate` resource defines configuration for an admin component, and a single broker
+(with future settings might be replicas for instance). Settings for these components can be tuned
+through this resource.
+
+```
+apiVersion: enmasse.io/v1alpha1
+kind: BrokeredInfraTemplate
+metadata:
+  name: brokered-infra-0.21.2
+  namespace: somewhere
+spec:
+  version: 0.21.2
+  broker:
+    addressFullPolicy: PAGE
+    resources:
+      memory: 2Gi
+      disk: 50Gi
+  admin:
+    resources:
+      memory: 512Mi
+```
+
```

Comment:
So, is it possible to have more than one instance of the same "kind" but with different names? e.g. "small-brokered-infra-0.21.2" and "large-brokered-infra-0.21.2" ?  what does the namespace here represent?

Reply:
The namespace would typically be the namespace where the EnMasse infra-controller is deployed I think, so 'enmasse-infra' or whatever.

Yes, you can have multiple instances with different names. This is to support multiple versions but also different configurations. They can be handled the same way, except different versions can have some different meanings in what schema is supported. The 'version' attribute is mainly to be able to specify what schema one can expect to follow.